mesonbuild/compilers/c.py

```diff
@@ -0,0 +1,996 @@
+                    if b'_' + symbol_name in line:
+                        return True
+                    # Else, check if the non-underscored form is present
+                    elif symbol_name in line:
+                        return False
+        raise RuntimeError('BUG: {!r} check failed unexpectedly'.format(n))
+
+    def find_library(self, libname, env, extra_dirs):
+        # First try if we can just add the library as -l.
+        code = '''int main(int argc, char **argv) {
+    return 0;
+}
+        '''
+        if extra_dirs and isinstance(extra_dirs, str):
+            extra_dirs = [extra_dirs]
+        # Gcc + co seem to prefer builtin lib dirs to -L dirs.
+        # Only try to find std libs if no extra dirs specified.
+        if not extra_dirs:
+            args = ['-l' + libname]
+            if self.links(code, env, extra_args=args):
+                return args
+        # Not found? Try to find the library file itself.
+        extra_dirs += self.get_library_dirs()
+        suffixes = ['so', 'dylib', 'lib', 'dll', 'a']
+        for d in extra_dirs:
+            for suffix in suffixes:
+                trial = os.path.join(d, 'lib' + libname + '.' + suffix)
+                if os.path.isfile(trial):
+                    return trial
+                trial2 = os.path.join(d, libname + '.' + suffix)
+                if os.path.isfile(trial2):
+                    return trial2
+        return None
+
+    def thread_flags(self):
+        return ['-pthread']
+
+    def thread_link_flags(self):
+        return ['-pthread']
+
+    def has_multi_arguments(self, args, env):
+        return self.compiles('int i;\n', env, extra_args=args)
+
+class VisualStudioCCompiler(CCompiler):
+    std_warn_args = ['/W3']
+    std_opt_args = ['/O2']
+
+    def __init__(self, exelist, version, is_cross, exe_wrap):
+        CCompiler.__init__(self, exelist, version, is_cross, exe_wrap)
+        self.id = 'msvc'
+        # /showIncludes is needed for build dependency tracking in Ninja
+        # See: https://ninja-build.org/manual.html#_deps
+        self.always_args = ['/nologo', '/showIncludes']
+        self.warn_args = {'1': ['/W2'],
+                          '2': ['/W3'],
+                          '3': ['/W4']}
+        self.base_options = ['b_pch'] # FIXME add lto, pgo and the like
+
+    # Override CCompiler.get_always_args
+    def get_always_args(self):
+        return self.always_args
+
+    def get_linker_debug_crt_args(self):
+        """
+        Arguments needed to select a debug crt for the linker
+
+        Sometimes we need to manually select the CRT (C runtime) to use with
+        MSVC. One example is when trying to link with static libraries since
+        MSVC won't auto-select a CRT for us in that case and will error out
+        asking us to select one.
+        """
+        return ['/MDd']
+
+    def get_buildtype_args(self, buildtype):
+        return msvc_buildtype_args[buildtype]
```

Comment:
__[flake8]__
* [F405] 'msvc_buildtype_args' may be undefined, or defined from star imports: .compilers

<!-- Comment by [SideCI](https://sideci.com) -->